grpc-api/src/main/java/io/deephaven/grpc_api/session/SessionState.java

```diff
@@ -383,12 +385,21 @@ public void addOnCloseCallback(final Closeable onClose) {
 
     /**
      * Remove an on-close callback bound to the life of the session.
+     * <p />
+     * A common pattern to use this will be for an object to try to remove itself, and if it succeeds, to call its own
+     * {@link Closeable#close()}. If it fails, it can expect to have close() be called automatically.
      *
      * @param onClose the callback to no longer invoke at end-of-life
      * @return true iff the callback was removed
+     * @apiNote If this SessionState has already begun expiration processing, {@code onClose} will not be removed by
+     *          this method. This means that if {@code onClose} was previously added and not removed, it either has
+     *          already been invoked or will be invoked by the SessionState.
      */
     public boolean removeOnCloseCallback(final Closeable onClose) {
         synchronized (onCloseCallbacks) {
+            if (isExpired()) {
+                return false;
+            }
             return onCloseCallbacks.remove(onClose) != null;
```

Comment:
So, this isn't really a guard for against multiple threads trying to remove the closeable (and invoke the close themselves) - this is a guard against *re-entrancy* (the synchronize already provides multi-threaded safety support). I feel like we are missing some context if we don't add a comment a "re-entrancy" comment above isExpired().

I'm happy to approve this, pending above; but I think there is room for a follow up where SimpleReferenceManager interface could be improved - at least with how we are trying to use it from SessionState:

```java
        synchronized (onCloseCallbacks) {
            onCloseCallbacks.forEach((ref, callback) -> {
                ...
                callback.close();
                ...
            });
            onCloseCallbacks.clear();
        }
```

The above is clumsy IMO, and if we could remove the callback before calling close (instead of afterwards during the clear) there wouldn't be any need for a re-entrancy check. I'm happy to file a ticket and tackle it later.

Reply:
Well, we already had a guard against different threads removing the closable, the synchronized block. This is specifically a guard against something removing an item while the same thread is already in the process of _iterating_ the list (at which point it will clear the list). Reentrancy is not itself a problem, except because we're storing "code" instead of objects - it is sort of a weird use of this to invoke arbitrary callbacks from inside the lock.

I think @rcaudy agrees that SimpleReferenceManager could do with some improvement.